### PR TITLE
feat(spanner): add samples for PG

### PR DESCRIPTION
**Pull request by rahul2393**
Adds samples and integration tests for PG features.

### CODE REVIEW

**Comment by rahul2393** [comment]
@codyoss Please help approve this

File: spanner/spanner_snippets/spanner/pg_spanner_create_database.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+		// includes database names in the CREATE DATABASE statement.
+		CreateStatement: `CREATE DATABASE "` + matches[2] + `"`,
+	})
+	if err != nil {
+		return err
+	}
+	if _, err := opCreate.Wait(ctx); err != nil {
+		return err
+	}
+	opUpdate, err := adminClient.UpdateDatabaseDdl(ctx, &adminpb.UpdateDatabaseDdlRequest{
```
Comment (olavloite) [inline spanner/spanner_snippets/spanner/pg_spanner_create_database.go:56]: nit: We should add a comment here that says something like:

```
Databases that are created with PostgreSQL dialect do not support extra DDL statements in the `CreateDatabase` call.
We must therefore execute these in a separate UpdateDatabaseDdl call after the database has been created.
```

File: spanner/spanner_snippets/spanner/pg_spanner_query_parameter.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+	client, err := spanner.NewClient(ctx, db)
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+
+	stmt := spanner.Statement{
+		SQL: `SELECT SingerId, FirstName, LastName
+              FROM Singers
+              WHERE LastName LIKE $1`,
```
Comment (olavloite) [inline spanner/spanner_snippets/spanner/pg_spanner_query_parameter.go:43]: nit: can we fix the indentation here?

**Review by olavloite — APPROVED** [review]

**Comment by telpirion** [comment]
This is a lot of region tags to add in a single PR. In the future, please break it up into multiple smaller PRs. That makes code review both faster & more efficient!

File: spanner/spanner_snippets/spanner/pg_spanner_connect_to_database.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	"context"
+	"io"
+
+	"cloud.google.com/go/spanner"
+	database "cloud.google.com/go/spanner/admin/database/apiv1"
+)
+
+// pgCreateClients shows how to create a database and admin client for a Spanner
+// PostgreSQL database.
+func pgCreateClients(w io.Writer, db string) error {
```
Comment (codyoss) [inline spanner/spanner_snippets/spanner/pg_spanner_connect_to_database.go:29]: Is this example needed? Other examples implicitly show this.

File: spanner/spanner_snippets/spanner/pg_spanner_create_database.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+	adminpb "google.golang.org/genproto/googleapis/spanner/admin/database/v1"
+)
+
+// pgCreateDatabase shows how to create a Spanner database that uses the
+// PostgreSQL dialect.
+func pgCreateDatabase(ctx context.Context, w io.Writer, db string) error {
+	// db := "projects/my-project/instances/my-instance/databases/my-database"
+	matches := regexp.MustCompile("^(.*)/databases/(.*)$").FindStringSubmatch(db)
+	if matches == nil || len(matches) != 3 {
+		return fmt.Errorf("Invalid database id %s", db)
```
Comment (codyoss) [inline spanner/spanner_snippets/spanner/pg_spanner_create_database.go:34]: nit: error strings should start with a lowercase letter: "invalid ..."

File: spanner/spanner_snippets/spanner/pg_spanner_order_nulls.go [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+		return err
+	}
+
+	client, err := spanner.NewClient(ctx, db)
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+
+	if _, err := client.Apply(context.Background(), []*spanner.Mutation{
```
Comment (codyoss) [inline spanner/spanner_snippets/spanner/pg_spanner_order_nulls.go:68]: Pass in the context you have already declared.

File: spanner/spanner_snippets/spanner/integration_test.go [hunk 6]
```diff
@@ -562,6 +562,300 @@ func TestCreateDatabaseWithDefaultLeaderSample(t *testing.T) {
...
+			user_name varchar(1024),
+			active    boolean
+		)`)
+	if dbCleanup != nil {
+		defer dbCleanup()
+	}
+	if err != nil {
+		t.Fatalf("failed to create test database: %v", err)
+	}
+	client, err := spanner.NewClient(context.Background(), dbName)
```
[the remark this answers is not in the document]
Comment (codyoss) [reply spanner/spanner_snippets/spanner/integration_test.go:791]: There are actually a lot of instances of this I notice of this /w spanner in other clients in this file. If you can update them all that would be great.

**Review by codyoss — CHANGES_REQUESTED** [review]
A few small things.

File: spanner/spanner_snippets/spanner/integration_test.go [hunk 7]
```diff
@@ -562,6 +562,300 @@ func TestCreateDatabaseWithDefaultLeaderSample(t *testing.T) {
...
+
+func TestPgDmlWithParameters(t *testing.T) {
+	_ = testutil.SystemTest(t)
+	t.Parallel()
+
+	_, dbName, cleanup := initTest(t, randomID())
+	defer cleanup()
+	dbCleanup, err := createTestPgDatabase(
+		dbName,
+		`CREATE TABLE Singers (
```
Comment (ryan-dayrit) [inline spanner/spanner_snippets/spanner/integration_test.go:629]: You can put this for maintainability/readability:
```
import _ "embed"
...
//go:embed testdata/create_table_singers.sql
var sqlCreateTableSingers string
```

File: spanner/spanner_snippets/spanner/pg_spanner_connect_to_database.go (hunk 3 above)
Comment (telpirion) [reply spanner/spanner_snippets/spanner/pg_spanner_connect_to_database.go:29]: Good point. You can nest this region tag inside one of the other samples.

File: spanner/spanner_snippets/spanner/pg_spanner_create_database.go [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+	}
+
+	adminClient, err := database.NewDatabaseAdminClient(ctx)
+	if err != nil {
+		return err
+	}
+	defer adminClient.Close()
+
+	opCreate, err := adminClient.CreateDatabase(ctx, &adminpb.CreateDatabaseRequest{
```
Comment (telpirion) [inline spanner/spanner_snippets/spanner/pg_spanner_create_database.go:43]: NIT: Consider removing struct instantiation to outside of the call to `CreateDatabase()`. It makes it easier for readers to easily parse what's going on.

File: spanner/spanner_snippets/spanner/pg_spanner_dml_with_parameters.go [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+              VALUES ($1, $2, $3), ($4, $5, $6)`,
+		// Use 'p1' to bind to the parameter with index 1.
+		Params: map[string]interface{}{
+			"p1": 1, "p2": "Alice", "p3": "Henderson",
+			"p4": 2, "p5": "Bruce", "p6": "Allison",
+		},
+	}
+	var updateCount int64
+	if _, err := client.ReadWriteTransaction(context.Background(), func(ctx context.Context, transaction *spanner.ReadWriteTransaction) error {
+		updateCount, err = transaction.Update(ctx, stmt)
```
Comment (telpirion) [inline spanner/spanner_snippets/spanner/pg_spanner_dml_with_parameters.go:50]: Interesting that the `stmt` variable is instantiated outside the scope of this function. I think this is probably better -- it's easier to read the values of `stmt`. I'll let other reviewers weigh in.

File: spanner/spanner_snippets/spanner/pg_spanner_information_schema.go [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+	// tables. The table_catalog is equal to the database name.
+
+	client, err := spanner.NewClient(ctx, db)
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+
+	// The `user_defined_...` columns are only available for PostgreSQL databases.
+	iter := client.Single().Query(ctx, spanner.Statement{SQL: `SELECT table_catalog, table_schema, table_name,
```
Comment (telpirion) [inline spanner/spanner_snippets/spanner/pg_spanner_information_schema.go:74]: Recommend moving the instantiation of `spanner.Statement` to outside of this method.

File: spanner/spanner_snippets/spanner/pg_spanner_interleaved_table.go [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+	if err != nil {
+		return err
+	}
+	defer adminClient.Close()
+
+	// The Spanner PostgreSQL dialect extends the PostgreSQL dialect with certain Spanner
+	// specific features, such as interleaved tables.
+	// See https://cloud.google.com/spanner/docs/postgresql/data-definition-language#create_table
+	// for the full CREATE TABLE syntax.
+	op, err := adminClient.UpdateDatabaseDdl(ctx, &adminpb.UpdateDatabaseDdlRequest{
```
Comment (telpirion) [inline spanner/spanner_snippets/spanner/pg_spanner_interleaved_table.go:45]: Recommend moving instantiation of `UpdateDatabaseDdlRequest` to outside of the call to `UpdateDatabaseDdl()`.

File: spanner/spanner_snippets/spanner/pg_spanner_numeric_data_type.go [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+
+	client, err := spanner.NewClient(ctx, db)
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+
+	var updateCount int64
+	// Insert a Venue using DML.
+	_, err = client.ReadWriteTransaction(ctx, func(ctx context.Context, transaction *spanner.ReadWriteTransaction) error {
```
Comment (telpirion) [inline spanner/spanner_snippets/spanner/pg_spanner_numeric_data_type.go:70]: Line 70 is also hard to parse. Suggest breaking it into multiple statements.

I liked how `spanner.Statement` was instantiated and then passed into the `func` in a previous sample.

File: spanner/spanner_snippets/spanner/pg_spanner_numeric_data_type.go [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+		})
+		return err
+	})
+	if err != nil {
+		return err
+	}
+	fmt.Fprintf(w, "Inserted %d venue(s)\n", updateCount)
+
+	// Insert a Venue with a NULL value for the Revenues column.
+	_, err = client.ReadWriteTransaction(ctx, func(ctx context.Context, transaction *spanner.ReadWriteTransaction) error {
```
Comment (telpirion) [inline spanner/spanner_snippets/spanner/pg_spanner_numeric_data_type.go:87]: Also a lot happening here. Consider breaking into multiple statements.

File: spanner/spanner_snippets/spanner/pg_spanner_numeric_data_type.go [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+			},
+		})
+		return err
+	})
+	if err != nil {
+		return err
+	}
+	fmt.Fprintf(w, "Inserted %d venue(s) with NULL revenues\n", updateCount)
+
+	// Insert a Venue with a NaN (Not a Number) value for the Revenues column.
```
Comment (telpirion) [inline spanner/spanner_snippets/spanner/pg_spanner_numeric_data_type.go:103]: This sample shows a lot of stuff! It looks like it demonstrates several things:

+ How to insert an actual number
+ How to insert a NULL value
+ How to insert a NaN value

I think that this will be hard to parse when published to cloud.google.com. I recommend providing multiple region tags so that each example (numeric, NULL, NaN) can be published independently.

**Review by telpirion — COMMENTED** [review]
Reviewed and commented; PTAL.

Some recurring observations:

+ A lot of structs are instantiated as literals inside of the methods where they're used. Although this makes for more terse code, it also makes it harder to read for customers. I recommend breaking instantiations outside of method calls so that they're easier to read & understand.
+ Similar to above, a lot of err checks happen as local variables within an if block. I think it's easier to read to have the err check follow the assignment of the err, but I think that's a preference thing.
+ Several sections have multiple things happening at once: a client method call, an error check, and a function declaration all within 1-2 lines. It's harder to parse what's going on with so much happening. I recommend breaking some of these lines into multiple statements.

**Comment by rahul2393** [comment]
@codyoss @telpirion Thanks for taking time to review this PR, I have updated the PR as per suggestions, please review again.

Sorry about the huge PR, will keep in mind to open smaller PR.

File: spanner/spanner_snippets/spanner/pg_spanner_numeric_data_type.go [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+
+	"cloud.google.com/go/spanner"
+	database "cloud.google.com/go/spanner/admin/database/apiv1"
+	"google.golang.org/api/iterator"
+	adminpb "google.golang.org/genproto/googleapis/spanner/admin/database/v1"
+)
+
+// pgNumericDataType shows how to work with the PostgreSQL NUMERIC/DECIMAL data
+// type on a Spanner PostgreSQL database.
+func pgNumericDataType(w io.Writer, db string) error {
```
[the remark this answers is not in the document]
Comment (rahul2393) [reply spanner/spanner_snippets/spanner/pg_spanner_numeric_data_type.go:34]: These sample tags were decided by @jin-jj, and all the client libraries were going to implement only those tags for PG samples. So only spanner_postgresql_numeric_data_type will be used for documentation rest will be ignored in this file.
